### PR TITLE
2d filter benchmark

### LINKED ISSUE #88: Benchmark 2D filter
In order to make a documented and informed decision on a good long-term solution to balance maintainability and performance #85 #73, we should compare the 2d filtering between
* original cython implementation
* numba implementation
* pure python implementation

both in terms of memory usage and speed.

### PR DESCRIPTION
**Pull request by alessandrofelder**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

## Description

**What is this PR**

- [ ] Bug fix
- [ ] Addition of a new feature
- [x] Other

**Why is this PR needed?**
Evaluate approaches to parallelise 2d filters

**What does this PR do?**

## References
closes #88 
see also #85

## How has this PR been tested?

n/a

## Is this a breaking change?

not going to be merged into main directly. Ultimate changes will not break API (they will replace underlying filtering code)

## Does this PR require an update to the documentation?

n/a

## Checklist:

- [ ] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: benchmarks/filter_2d.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from pyinstrument import Profiler
+
+from cellfinder_core.detect.filters.plane import TileProcessor
+from cellfinder_core.detect.filters.setup_filters import setup_tile_filtering
+
+# Use random 16-bit integer data for signal plane
+shape = (10000, 10000)
+
+signal_array_plane = np.random.random(shape)
+signal_array_plane = (signal_array_plane * 65535).astype(np.uint16)
```
Comment (dstansby) [inline benchmarks/filter_2d.py:11]:
```suggestion
signal_array_plane = np.random.randint(low=0, high=65535, size=shape, dtype=np.uint16)
```
Is a bit clearer (I discovered `randint` after writing the original benchmark)

File: benchmarks/filter_2d.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import numpy as np
+from pyinstrument import Profiler
```
[the remark this answers is not in the document]
Comment (adamltyson) [reply benchmarks/filter_2d.py:2]: Don't see why not, unless it comes with any installation headaches.

**Review by adamltyson — APPROVED** [review]

File: benchmarks/filter_2d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dstansby) [reply benchmarks/filter_2d.py:2]: Nope, it's lightweight with no dependencies itself.

File: src/cellfinder_core/detect/filters/plane/base_tile_filter.py [hunk 3]
```diff
@@ -1,106 +1,91 @@
...
-    cpdef is_low_average(self):  # TODO: move to OutOfBrainTileFilter
-        cdef bint is_low
-        cdef double avg = 0
-        cdef uint x, y
-        for x in range(self._tile.shape[0]):
-            for y in range(self._tile.shape[1]):
-                avg += self._tile[x, y]
-        avg /= self._tile.shape[0] * self._tile.shape[1]
+    @jit
+    def is_low_average(self, tile):  # TODO: move to OutOfBrainTileFilter
```
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/plane/base_tile_filter.py:40]: it does (in its current form at least) depend on `self` and seems to update `self.keep` - do you think it's worth restructuring as part of this PR?

[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/detect/filters/plane/base_tile_filter.py:40]: Wooops, don't know how I missed that... can definitely keep this here then 👍

**Review by dstansby — APPROVED** [review]
Looks like linting needs fixing, but feel free to self-merge when that's fixed.

**Pull request merged**